keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,659 @@
+don't want to start my workload if the pods cannot be placed within the same
+rack.
+
+#### Story 2
+
+As an ML researcher I run AI training workloads which require communicating
+moderate amounts of data between pods. I would like to optimize the runtime of
+the workloads, by placing the pods as close to another as possible, ideally
+within a single rack, but I'm ok running the workload on nodes scattered across
+datacenter if placing all pods within the same hierarchy rack is not possible at
+the given time.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Integration support
+
+In the alpha iteration we aim to support Job & JobSet. If other integrations
+follow naturally we may as well support them in alpha.
+
+##### Job
+
+**Example**:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  namespace: tas-example-job
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+  annotations:
+    kueue.x-k8s.io/require-topology: cloud.provider.com/topology-rack
+spec:
+  parallelism: 10
+  completions: 10
+  template:
+    spec:
+      containers:
+      - name: worker
+        image: gcr.io/k8s-staging-perf-tests/sleep:v0.1.0
+        args: ["300s"]
+        resources:
+          requests:
+            cpu: "4"
+      restartPolicy: Never
+```
+
+In this example we indicate that all Pods created by the Job should be contained
+within the same "rack".
+
+##### JobSet
+
+One complication we noticed for JobSet is that the proposed design assumes
+injecting the dedicated scheduling gate, called `kueue.x-k8s.io/topology` into
+the PodTemplate. This required a JobSet fix which is released in 0.6
+(see [PR](https://github.com/kubernetes-sigs/jobset/pull/623)).
+
+**Example**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: tas-example-jobset
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+  annotations:
+    kueue.x-k8s.io/prefer-topology: cloud.provider.com/topology-rack
+spec:
+  replicatedJobs:
+  - name: leader
+    replicas: 1
+    template:
+      spec:
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: leader
+              image: gcr.io/k8s-staging-perf-tests/sleep:v0.1.0
+              args: ["300s"]
+  - name: workers
+    replicas: 2
+    template:
+      spec:
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: gcr.io/k8s-staging-perf-tests/sleep:v0.1.0
+              args: ["100s"]
+```
+
+In this example we say that all PodSets corresponding to instances of
+replicated Jobs (one leader and two workers) prefer the "rack" topology domain.
+
+Note that, there is no option to select a different topology domain per replica
+type, or to indicate the topology domain for the collection of Jobs. We will
+consider this for [Beta](#beta) based on users feedback.
+
+#### Support for the "auto" mode
```

Comment:
I'm wondering if we should actually start with the support for "auto" and defer specifying a level. I think this would improve the UX, and still cover the main use-case. We would also not introduce the duality between the "auto" and the specific value. Maybe for specifying a level, if ever needed, we can have a dedicated annotation "prefer-level". WDYT @mwielgus @alculquicondor ?

Reply:
I synced with @mwielgus and the idea is to not introduce auto in alpha. The reason is that this is a complex feature so we want users to learn about the intricacies of the feature. We will consider simplifying in the future.